primitives/core/src/lib.rs

```diff
@@ -465,3 +465,55 @@ macro_rules! impl_maybe_marker {
 // The maximum possible allocation size was chosen rather arbitrary, 32 MiB should be enough for
 // everybody.
 pub const MAX_POSSIBLE_ALLOCATION: u32 = 33554432; // 2^25 bytes, 32 MiB
+
+/// Executes the wrapped closure on drop.
+///
+/// Should be used together with the [`defer`] function.
+pub struct DeferGuard<F: FnOnce()>(Option<F>);
+
+impl<F: FnOnce()> Drop for DeferGuard<F> {
+	fn drop(&mut self) {
+		self.0.take().map(|f| f());
+	}
+}
+
+/// Construct a [`DeferGuard`] that will execute the given function on drop.
+///
+/// Multiple calls to [`defer`] will execute the closures in reverse order.
+///
+/// NOTE: This could also be implemented as macro but then it would not work anymore
+/// when called multiple times in the same scope since it needs to create a variable.
```

Comment:
Not sure how this helps. You should add some doc tests to illustrate how to use. That would help much more. Imo.

Reply:
Yep done